cm_ovirt_vm_creator.py

```diff
@@ -143,47 +143,103 @@ def construct_search_by_name_query(node_name):
-                time.sleep(args.sleep_between_iterations)
-                vm = vm_service.get()
-                print("vm.status = %s" % (vm.status), file=sys.stderr)
-                if vm.status == types.VmStatus.DOWN:
-                    break
-                counter += 1
-            pub_sshkey = os.environ[args.pub_sshkey]
-            vm_service.start(use_cloud_init=True,
-                             vm=types.Vm(initialization=types.Initialization(authorized_ssh_keys=pub_sshkey)))
-            counter = 1
-            while counter < args.num_of_iterations:
-                time.sleep(args.sleep_between_iterations)
-                vm = vm_service.get()
-                print("vm.status = %s, vm.fqdn= '%s'" % (vm.status, vm.fqdn), file=sys.stderr)
-                if vm.status == types.VmStatus.UP:
-                    break
-                counter += 1
-
-            if vm.status != types.VmStatus.UP:
-                print("ERROR - VM {0} still not up after {1} retries".format(node, args.num_of_iterations), file=sys.stderr)
-                sys.exit(-1)
-            else:
-                vm_dict[node] = vm_service
+            to_create.append(node)
+
+    # Create the VM in "blocks"
+    for block in chunks(to_create, args.block_size):
+        block_futures = []
+        for node in block:
+            vm_future = vms_service.add(types.Vm(name=node,
+                                                 cluster=types.Cluster(name=args.ovirt_cluster),
+                                                 template=types.Template(name=args.ovirt_template)), wait=False)
+            block_futures.append((node, vm_future))
+        # wait for all the VMs from this block to be created
+        for node_name, future_vm in block_futures:
+            vm = future_vm.wait()
+            vm_dict[node_name] = vms_service.vm_service(vm.id)
+        # sleep before the next block
+        time.sleep(args.sleep_between_iterations)
+
+    # Start each VM when it's created, but try to batch the calls
+    counter = 1
+    started = set()
+    pub_sshkey = os.environ[args.pub_sshkey]
+    # number of attempts is bigger here because it's not attempts per VM
+    # like in the other nodes.
+    while counter < args.num_of_iterations * len(cluster_nodes):
+        start_futures = []
+        for node_name, vm_service in vm_dict.items():
+            if node_name in started:
+                continue
+            vm = vm_service.get()
+            print("%s: vm.status = %s" % (node_name, vm.status), file=sys.stderr)
+            if vm.status == types.VmStatus.DOWN:
+                print("%s: starting" % (node_name), file=sys.stderr)
+                future = vm_service.start(use_cloud_init=True, wait=False,
+                                          vm=types.Vm(initialization=types.Initialization(authorized_ssh_keys=pub_sshkey)))
+                start_futures.append(future)
+                started.add(node_name)
```

Comment:
Is this right?  I think `started` should only contain the UP ones; if you `.add()` here too, I think you'll always end up with `set(started) == set(cluster_nodes)`, even if they're still starting...

Reply:
OTOH, if you don't `started.add()` you risk invoking `vm_service.start()` multiple times on same VM.
OK, I see below you wait again to really be up.
So this looks right :+1:, but consider renaming `started` to `starting`.

Reply:
I named it "started" as in "we started the VM", meaning every VM on which we called `.start()`.

But yeah I see why this might be a bit confusingly named.